### PR TITLE
Depth2 items

### PR DESCRIPTION
**Pull request by GeoffThorpeFT**
Allow read/write of attributes of related items

### CODE REVIEW

**Comment by GeoffThorpeFT** [comment]
Please note that this pull request changes the format of the json output for items.
Not only does it show attributes of related items - its also prefixes the key value with 'dataItemID'. Before we only output the key, not the name of the key field.

File: cmdbv2api/views.py [hunk 1]
```diff
@@ -1410,18 +1410,26 @@ def itemContent(requestMethod, requestData, typeID, itemID):
...
-                        checkReferenceData(requestMethod, "item", DataItemSerializer, dictQuery)
+                        oitemID = oitem.get('dataItemID', None)
+                        if oitemID:
+                            # check the related item - create if necessary
+                            dictQuery = {'dataTypeID':DataType.objects.get(ID=itype), 'dataItemID': oitemID}
+                            checkReferenceData(requestMethod, "item", DataItemSerializer, dictQuery)
+
+                            # as we are dealing with a related item with attributes we can now popukate that item with its attributes
+                            oitemattrs = deepcopy(oitem)
+                            oitemattrs.pop('dataItemID')
```
Comment (lucas42) [inline cmdbv2api/views.py]: Might be worth considering leaving this in.  I know it's duplicated by the key, but it'd make it consistent with the top-level item, potentially making it easier for clients to use the data in templates etc.

(P.S. The more we use this, the more I can't wait till it gets renamed 😝 )

File: cmdbv2api/serializers.py [hunk 2]
```diff
@@ -60,6 +60,36 @@ class Meta:
...
+        fieldarray = outputfields.split(',')
+
+        # now squash attribute array into list of name/value pairs
+        attributes = data.pop('attributes', '')
+        # loop thorugh array to just store name/value pairs
+        for attribute in attributes:
+            if outputfields == '__ALL__':
+                data.update({attribute['attributeType']:attribute['value']})
+            if attribute['attributeType'] in fieldarray:
+                data.update({attribute['attributeType']:attribute['value']})
```
Comment (lucas42) [inline cmdbv2api/serializers.py]: Might be clearer to turn the 2 if statements into one with an `or`, so you don't have to repeat this line twice.

File: cmdbv2api/models.py [hunk 3]
```diff
@@ -139,6 +139,7 @@ class DataType(models.Model):
     minLength = models.IntegerField(blank=True, null=True)
     maxLength = models.IntegerField(blank=True, null=True)
     convention = models.CharField(max_length=255, blank=True)
+    outputFields = models.CharField(max_length=255, blank=True)
```
Comment (GeoffThorpeFT) [inline cmdbv2api/models.py]: Which native format would you like? Bearing in mind its currently just a text field.

File: cmdbv2api/views.py (hunk 1 above)
Comment (GeoffThorpeFT) [inline cmdbv2api/views.py]: I think you may be assuming dataItemID is being popped to stop it appearing in the output; its actually being popped to stop it appearing in the attributes of the linked item (the old error of attributes named after keys).

File: cmdbv2api/models.py (hunk 3 above)
Comment (lucas42) [inline cmdbv2api/models.py]: I meant not a text field.  But I don't think django has any fields for arrays or lists.  And it's probably too much effort to model them as a separate type, right?  
In which case just make it clearer in the definition what representation you're using.

File: cmdbv2api/views.py (hunk 1 above)
Comment (lucas42) [inline cmdbv2api/views.py]: Ah, sorry.  Yes, that makes sense.

File: cmdbv2api/models.py (hunk 3 above)
Comment (GeoffThorpeFT) [inline cmdbv2api/models.py]: That would be my approach - suitable documentation in the doc text plus something similar on an gui we create

File: cmdbv2api/views.py (hunk 1 above)
Comment (GeoffThorpeFT) [inline cmdbv2api/views.py]: but I fully support the need to rename the field!

File: cmdbv2api/serializers.py [hunk 4]
```diff
@@ -73,24 +103,41 @@ class Meta:
...
         super(DataItemSerializer, self).__init__(*args, **kwargs)
 
 
     def to_representation(self, obj):
         data = super(DataItemSerializer, self).to_representation(obj)
 
+        # we need to get identification info for related objects (so we can show more than a key)
+        datatypes = self.context.get('datatypes', {})
+        relatedItemIDs = []
+        relatedJSON = {}
```
Comment (GeoffThorpeFT) [inline cmdbv2api/serializers.py]: I think it is JSON - its the result of a call to the data method of the serializer  (see line 133)

Comment (GeoffThorpeFT) [inline cmdbv2api/serializers.py]: .. or was your statement aligned to the fact that the data method is actually returning a dict which the django response function converts to json when the view returns.

**Comment by GeoffThorpeFT** [comment]
Re Migration of existing users - I think that is only the couple of early UI apps we wrote: dewey and contact-manager.  Both of which I intend to update (as they may also now conflict with the system registry field names)

**Comment by lucas42** [comment]
👍

**Pull request merged**